### ISSUE TITLE
It's not possible to choose a country 

### ISSUE THREAD

**Issue by kouros17** [manjaro/pamac#366]
After the last changes of **[pacman-mirrors.conf](https://forum.manjaro.org/t/pacman-mirrors-release-of-version-4-5b1/34446)**, it's not possible to choose a country other than "worldwide" (pamac 6.2.2-1)

![2017-11-14 14-03-53](https://user-images.githubusercontent.com/27904795/32779066-074b2518-c945-11e7-861a-53acdd6ce0e2.jpg)

**Comment by philmmanjaro** [comment]
@kouros17: which **pacman-mirrors** do you use? @fhdk: did we missed some?

![pamac-v6 2 2](https://user-images.githubusercontent.com/5050022/32779331-54f173ce-c90b-11e7-8a89-335f22b944f3.png)

**Comment by kouros17** [comment]
I use pacman-mirrors 4.5b1-2. As

**Comment by fhdk** [comment]
I will look at IT tomorrow. 

It is without doubt a question of a too limited custom-mirror pool.

I do not know how pamac does its thing.

Have you tried to have a look at pamac output window or done the command in terminal?

I am going to sleep now.

**Comment by fhdk** [comment]
```
~ >>> pacman-mirrors -v                                                                                  
Version 4.5b1
```
Pamac v 6.2.2-1

![pamac-country-selection](https://user-images.githubusercontent.com/22748698/32832773-ef8a6e4e-c9fc-11e7-861e-fa0a4a1edf28.png)

As said before I don't know the inner  workings of pamac but pacman - I know inside out.

And with the change in b1 - the `OnlyCountry = ` has been deprecated and pacman-mirrors verifies the existence of a `custom-mirrors.json` and will generate your mirrorlist from that pool.

If your mirror pool does not contain any mirrors which is up-to-date no changes will be made to the mirrorlist.

```
~ >>> sudo pacman-mirrors -c Denmark                                                                     
[sudo] password for fh: 
.: INFO Downloading mirrors from repo.manjaro.org

.: INFO User generated mirror list
--------------------------
.: INFO Custom mirror file saved: /var/lib/pacman-mirrors/custom-mirrors.json
.: INFO Using default mirror file
.: INFO Querying mirrors - This may take some time
   0.076 Denmark        : https://mirrors.dotsrc.org/manjaro/
   0.072 Denmark        : https://www.uex.dk/public/manjaro/
.: INFO Writing mirror list
   Denmark         : https://www.uex.dk/public/manjaro/unstable
   Denmark         : https://mirrors.dotsrc.org/manjaro/unstable
.: INFO Mirror list generated and saved to: /etc/pacman.d/mirrorlist
```
This command generated a custom-mirrors.json which made no change to the pamac country listing. Using the refresh mirrorlist in pamac does work well too

![after-custom-mirror-pool-generation](https://user-images.githubusercontent.com/22748698/32833170-43a5c540-c9fe-11e7-923e-365233ce3811.png)

**Comment by philmmanjaro** [comment]
@fhdk: thx for your detailed explanation. We have to see if we have here an error.

**Comment by fhdk** [comment]
I have tried to force the issue but seems I can't - I remember though I have seen it during development. So in theory it could exist in pre b1.

I made a series of screen prints trying to force the issue
<hr/>
* loading the default mirror pool
* selecting mirrors with only http
* then setting pacman-mirrors.conf to accept only https

![pacman-reset-all-select-only-http-mirrors-and-then-change-to-https](https://user-images.githubusercontent.com/22748698/32834153-ce9f7daa-ca01-11e7-9fca-1926c79ce46c.png)

![pacman-mirrors-id-select-only-http](https://user-images.githubusercontent.com/22748698/32834177-e419bbe6-ca01-11e7-8210-9c01f793290c.png)
<hr/>
pamac refresh mirrorlist
![pamac-refresh-mirrors-with-https-in-config](https://user-images.githubusercontent.com/22748698/32834254-35e56b64-ca02-11e7-9fc2-700d0fc22425.png)
<hr />
pamac refresh database - seems done twice (maybe first mirror did not answer - I don't know)
![pamac-refresh-database-action-seems-done-twice](https://user-images.githubusercontent.com/22748698/32834299-6c50dde6-ca02-11e7-8034-07c01d886806.png)

**Comment by fhdk** [comment]
@guinux @philmmanjaro @kouros17 

**My missing response**
I have been busy with other things and have not read the comment thread on the forum until now.
Now I know exactly what is causing the missing selection.
I assume that pamac writes the users country selection to the `OnlyCountry = ` config option.

**Why it happens**
The problem arises because in 4.5b1 this exact option is not used by pacman. It is no problem to roll back - but based on the feedback - verbal or likes - it seemed a good solution just to check for the existence of the custom-mirrors.json.

**Proper solution**
Please let me know if I should drop back to 4.4.1 functionality?

**Comment by guinux** [comment]
Yes, pamac reads/writes `OnlyCountry` option.
I can adopt new pacman-mirrors and use `-c` option to generate mirrorlist with one country but how can pamac have access to previous choosen country ?

**Comment by fhdk** [comment]
I did not realize that pamac was using that entry in the config.

The users custom mirror pool created with `-c` is saved in `/var/lib/pacman-mirrors/custom-mirrors.json` in the exact same format as the `status.json` in the same location.

From a separation of concerns POW it would be preferable to have the data files as the exchange point.

But I do see two options here 

- the first one requiring @guinux to build one or more functions to parse json 
- the other is me writing to `/etc/pamac.conf` to inform pamac of the users preferred countries.

1. pamac [parses the json files](https://valadoc.org/json-glib-1.0/Json.Parser.html)
   * check **if** `custom-mirrors.json` exist - use it **else** use `status.json`.
2. pacman-mirrors check if pamac.conf is present (pamac installed) and writes the list of countries as an entry in pamac.conf

```
### Pamac configuration file

## When removing a package, also remove those dependencies
## that are not required by other packages (recurse option):
#RemoveUnrequiredDeps

## How often to check for updates, value in hours (0 to disable):
RefreshPeriod = 6

## When there are no updates available, hide the tray icon:
#NoUpdateHideIcon

## Allow Pamac to search and install packages from AUR:
EnableAUR

## When AUR support is enabled check for updates from AUR:
CheckAURUpdates

## AUR build directory:
BuildDirectory = /tmp

## Number of versions of each package to keep in the cache:
KeepNumPackages = 3

## Remove only the versions of uninstalled packages when clean cache:
#OnlyRmUninstalled

## Comma separated list of countries from `/var/lib/pacman-mirrors/custom-mirrors.json`
# CountryList = 
```

**Comment by fhdk** [comment]
@guinux @philmmanjaro 
I would very much like to close this one but which approach do we take?

* Let pacman-mirrors modify pamac.conf?
* Learning pamac to parse the json for countries?

**Comment by philmmanjaro** [comment]
@fhdk I think @guinux will find a proper way. If it is needed that **pacman-mirrors** should touch the **pamac.conf** we let you know.

**Comment by fhdk** [comment]
@philmmanjaro @guinux 
Ok - great - but I we should make sure 4.5b1 does not move to testing or stable. That is - remove it from unstable - replacing it with 4.4 - I will make an post explaining the reason to the incompatibility with pamac.

When a solution which works for pamac is in place we put the 4.5 on the table again.

**Comment by guinux** [comment]
I want to keep pamac as simple as possible for user. This means that I would like to keep the current view in pamac with only one country selected.
Of course I could parse `custom-mirrors.json` but this file can be created with multiple countries, and without all mirrors of a country if `--interactive` option is used.

I would suggest for pacman-mirrors to:
- keep `OnlyCountry` variable in config file and change it to `Custom` only if **multiple countries** are given or if `--interactive` option is used.
- create `custom-mirrors.json` file only if it's required to change `OnlyCountry` variable to `Custom`.

This way we can have a simple way to use pacman-mirrors for beginners with one country in `OnlyCountry` and for advanced users which used the command line for more control, this will be changed to `Custom`.

**Comment by fhdk** [comment]
@philmmanjaro @guinux 
I get your idea but I still think it's a bad habit to write other apps config.
This specific issue a really good argument for an app **not** writing to another app's config.

The specific reason for deprecating the `OnlyCountry` was because it was confusing. As can be read in the forum reference in OP.


You don't even have to parse json with this approach. You still keep it simple for your users - just let them select the country and let `pacman-mirrors` create the `custom-mirrors.json` with the single chosen country.

* Create an entry in `pamac.conf` to hold a users Pamac selected country.
* If entry is empty `Worldwide` is assumed.
   * When the user selects a country write it to `pamac.conf`.
   * Based on the users selection 
      * COUNTRY: call `pacman-mirrors -c COUNTRY`
      * Worldwide: call `pacman-mirrors -call`.

* You know it has changed if `custom-mirrors.json` does not exist.
* A user would and should not expect a change made by pacman-mirrors to be reflected in pamac.

As long as you let pacman-mirrors create the mirror pool you are good.

Another idea I can come up with is to create a similar output to `-l` which could be `-sc` which would output the selected countries from the configuration. You can then make your decisions based on that output.

**Comment by guinux** [comment]
@fhdk Your idea is near from one I had and is good as long as, like you said:

> A user would and should not expect a change made by pacman-mirrors to be reflected in pamac

I can go with it if we all agree.
@philmmanjaro what's your opinion ?

**Comment by philmmanjaro** [comment]
@guinux: I recommend the same as @fhdk. Use pamac.conf for saving the country and only trigger **pacman-mirrors** with given switches. This will guarantee no mixup within the apps.

**Comment by guinux** [comment]
Ok, I will go for it.

**Comment by fhdk** [comment]
@guinux 
When pacman-mirrors launch it checks if the original mirrorfile in /usr/share has changed. It does that my checksumming the two files to see if the existing file needs updating.

When your user has selected a country - and pacman-mirrors has done its thing - you could checksum the generated `custom-mirrors.json` and save that checksum beside your country selection.

When pamac is launched you can check for existance of custom-mirrors.json - compare checksum if exist - and let the user know that you know the pacman-mirrors info has changed and you can ask questions or default or  .....

**Comment by guinux** [comment]
Can you had a function to pacman-mirrors which just do the checksum and return if the mirrorlist needs to be updated? It can help me.

**Comment by fhdk** [comment]
I use a python library `filecmp`

When your user has selected and pacman returns you could do
```
md5sum /var/lib/pacman-mirrors/custom-mirrors.json                                                 
9279c999cf9548fd3065fd3d2a1c627a  /var/lib/pacman-mirrors/custom-mirrors.json
```
Store the checksum in a key in your conf.
When user loads the preferences you could do the same command and compare the result to your stored value. If not equal then user has used pacman-mirrors to change the config and you can act accordingly.

**Comment by fhdk** [comment]
@guinux 
> Can you had a function to pacman-mirrors which just do the checksum and return if the mirrorlist needs to be updated? It can help me.

How can  I do that?

**Comment by fhdk** [comment]
@guinux @philmmanjaro 

This [post](https://forum.manjaro.org/t/pamac-mirror-select-thingy-not-working-post-includes-a-video/42774/7) draw my attention to this old issue.

I have been thinking on the country selection issue.

As of v4.7.6 you ask pacman-mirrors about version, countries and branch.

    pacman-mirrors -v
    pacman-mirrors -l
    pacman-mirrors -aG

Pamac can easily set a country by using the `--country` argument to pacman-mirrors.

    pacman-mirrors -c {country[, country ...]}

Pamac can store that in its own config. But the mirror list result might not be as expected as the mirrors might not be up-to-date and pacman-mirrors only uses up-to-date mirrors. 

For this to be useful in terms of Pamac functionality, pacman-mirrors must be enhanced with an argument which allows for mirrors not up-to-date.

Also I can enhance pacman-mirrors by adding an argument which returns the countries currently in the custom-mirrors.json.

Let me know what you think. 

I have added some issues 
https://github.com/manjaro/pacman-mirrors/issues/125
https://github.com/manjaro/pacman-mirrors/issues/127
https://github.com/manjaro/pacman-mirrors/issues/128

**Comment by fhdk** [comment]
@guinux 

I have build a dev package and made it available [here](https://www.uex.dk/public/dev-pkg)

The pacman-mirrors in the package has a new argument `-lc/--country-config` which can be called without root.

It is similar to the `-l/--list/--country-list` and upon execution it will return the countries listed in `custom-mirrors.json` if it exists otherwise nothing.

Let me know if it works for you.

**Comment by guinux** [comment]
What pamac needs is to have access to all available countries and the custom choosen country.

**Comment by fhdk** [comment]
@guinux @philmmanjaro 
https://github.com/manjaro/pacman-mirrors/releases/tag/4.9.0

**Comment by fhdk** [comment]
@guinux 

Usual country list
* `pacman-mirrors -l`

Custom country list
* `pacman-mirrors -lc`

**Comment by philmmanjaro** [comment]
So with `-l` we get the total list of countries available and with `-c` you can define one. I don't think we will use the interactive one. @guinux does this functionality help **pamac**?

**Comment by guinux** [comment]
I will use:
`pacman-mirrors -l` to display all available countries
`pacman-mirrors -lc` to retrieve the previous choosen country
`pacman-mirrors -c` with the new choosen country

**Comment by fhdk** [comment]
>`pacman-mirrors -lc` to retrieve the previous choosen country

Please be aware of possible return values.

* empty list
* one country
* multiple countries

**Comment by guinux** [comment]
Changes adopted :)
See  0bfe35e

**Comment by kouros17** [comment]
Fixed in **6.2.5-4**
